File: stages/diagram-recognition/src/main/java/edu/kit/kastel/mcse/ardoco/erid/diagramrecognition/DiagramRecognitionMock.java
/* Licensed under MIT 2023. */
package edu.kit.kastel.mcse.ardoco.erid.diagramrecognition;

import java.util.List;
import java.util.SortedMap;

import edu.kit.kastel.mcse.ardoco.core.api.diagramrecognition.DiagramRecognitionState;
import edu.kit.kastel.mcse.ardoco.core.common.util.wordsim.UnicodeCharacterMatchFunctions;
import edu.kit.kastel.mcse.ardoco.core.common.util.wordsim.WordSimUtils;
import edu.kit.kastel.mcse.ardoco.core.data.DataRepository;
import edu.kit.kastel.mcse.ardoco.core.pipeline.ExecutionStage;
import edu.kit.kastel.mcse.ardoco.erid.diagramrecognition.agents.DiagramDisambiguationAgent;
import edu.kit.kastel.mcse.ardoco.erid.diagramrecognition.agents.DiagramReferenceAgent;
import edu.kit.kastel.mcse.ardoco.lissa.DiagramRecognitionStateImpl;
import edu.kit.kastel.mcse.ardoco.tests.eval.GoldStandardDiagrams;

/**
 * This stage is responsible for mocking the {@link edu.kit.kastel.mcse.ardoco.lissa.DiagramRecognition} stage. It populates the {@link DiagramRecognitionState}
 * using the {@link GoldStandardDiagrams} gold standard.
 */
@SuppressWarnings("java:S110")
public class DiagramRecognitionMock extends ExecutionStage {
    private final WordSimUtils wordSimUtils;
    private final GoldStandardDiagrams goldStandardProject;

    public DiagramRecognitionMock(GoldStandardDiagrams goldStandardProject, SortedMap<String, String> additionalConfigs, DataRepository dataRepository) {
        super(List.of(new DiagramDisambiguationAgent(dataRepository), new DiagramReferenceAgent(dataRepository)), DiagramRecognitionMock.class.getSimpleName(),
                dataRepository, additionalConfigs);
        this.goldStandardProject = goldStandardProject;
        this.wordSimUtils = getMetaData().getWordSimUtils();
    }

    @Override
    protected void initializeState() {
        logger.info("Creating DiagramRecognitionMock State");
        var diagramRecognitionState = new DiagramRecognitionStateImpl();
        var diagrams = goldStandardProject.getDiagramsGoldStandard();
        for (var diagram : diagrams) {
            logger.debug("Loaded Diagram {}", diagram.getResourceName());
            diagramRecognitionState.addDiagram(diagram);
        }
        getDataRepository().addData(DiagramRecognitionState.ID, diagramRecognitionState);
    }

    private UnicodeCharacterMatchFunctions previousCharacterMatchFunction;

    @Override
    protected void before() {
        super.before();
        previousCharacterMatchFunction = wordSimUtils.getCharacterMatchFunction();
        wordSimUtils.setCharacterMatchFunction(UnicodeCharacterMatchFunctions.EQUAL_OR_HOMOGLYPH);
    }

    @Override
    protected void after() {
        wordSimUtils.setCharacterMatchFunction(previousCharacterMatchFunction);
        super.after();
    }
}

File: stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/erid/diagramrecognition/DiagramRecognitionMock.kt
package edu.kit.kastel.mcse.ardoco.erid.diagramrecognition

import edu.kit.kastel.mcse.ardoco.core.api.diagramrecognition.DiagramRecognitionState
import edu.kit.kastel.mcse.ardoco.core.common.util.wordsim.UnicodeCharacterMatchFunctions
import edu.kit.kastel.mcse.ardoco.core.common.util.wordsim.WordSimUtils
import edu.kit.kastel.mcse.ardoco.core.data.DataRepository
import edu.kit.kastel.mcse.ardoco.core.pipeline.ExecutionStage
import edu.kit.kastel.mcse.ardoco.erid.diagramrecognition.agents.DiagramDisambiguationAgent
import edu.kit.kastel.mcse.ardoco.erid.diagramrecognition.agents.DiagramReferenceAgent
import edu.kit.kastel.mcse.ardoco.lissa.DiagramRecognitionStateImpl
import edu.kit.kastel.mcse.ardoco.tests.eval.GoldStandardDiagrams
import java.util.SortedMap

/**
 * This stage is responsible for mocking the [edu.kit.kastel.mcse.ardoco.lissa.DiagramRecognition] stage. It populates the [DiagramRecognitionState]
 * using the [GoldStandardDiagrams] gold standard.
 */
class DiagramRecognitionMock(
    private val goldStandardProject: GoldStandardDiagrams?,
    additionalConfigs: SortedMap<String, String>?,
    dataRepository: DataRepository?
) :
    ExecutionStage(
            listOf(DiagramDisambiguationAgent(dataRepository), DiagramReferenceAgent(dataRepository)),
            DiagramRecognitionMock::class.java.getSimpleName(),
            dataRepository,
            additionalConfigs
        ) {
    private val wordSimUtils: WordSimUtils = getDataRepository().globalConfiguration.wordSimUtils

    override fun initializeState() {
        logger.info("Creating DiagramRecognitionMock State")
        val diagramRecognitionState = DiagramRecognitionStateImpl()
        val diagrams = goldStandardProject?.getDiagramsGoldStandard() ?: listOf()
        for (diagram in diagrams) {
            logger.debug("Loaded Diagram {}", diagram.resourceName)
            diagramRecognitionState.addDiagram(diagram)
        }
        getDataRepository().addData(DiagramRecognitionState.ID, diagramRecognitionState)
    }

    private var previousCharacterMatchFunction: UnicodeCharacterMatchFunctions? = null

    override fun before() {
        super.before()
        previousCharacterMatchFunction = wordSimUtils.characterMatchFunction
        wordSimUtils.characterMatchFunction = UnicodeCharacterMatchFunctions.EQUAL_OR_HOMOGLYPH
    }

    override fun after() {
        wordSimUtils.characterMatchFunction = previousCharacterMatchFunction
        super.after()
    }
}
